Third party API design document (#10)
pace-neutrons/pace-developers: optimisation/design/Third_Party_API_Design.md

# An Application Programming Interface for user defined model functions in PACE
**2020-05-15**

# Scope

This documents details the required function signatures for user defined sample scattering models
and how they interface with Horace, using either Matlab, Python, C, C++ or Fortran.

# Motivation

Analysis of inelastic neutron scattering data almost always requires a parameterised model of the measured scattering intensities.
Horace, as currently implemented, requires this model to be in the form of a Matlab function.
This function is required to either 

1. Take a 4D vector of momentum-energy ![\mathbb{Q} \equiv (\mathb{Q},E)] and yield the scattering function ![S(\mathbb{Q})].

> 2. Take the 3D vector of momentum transfer ![\mathbf{Q}] only and yield ![N] sets of vectors ![(E_n, S(\mathbf{Q}, E_n))] where ![N] is the number of dispersion branches.

Reviewer: I would propose a third variant which takes **Q** and returns E<sub>n</sub>, S(**Q**,E<sub>n</sub>), and γ(**Q**,E<sub>n</sub>); where γ(**Q**,E<sub>n</sub>) *is* the intrinsic energy width of the n<sup>th</sup> mode at **Q**.

Author: This text is about the current implementation...

For type 2, the ![N] sets of vectors are contained in a cell array
and a helper function `disp2sqw` is used to obtain ![S(\mathbb{Q})]
by convolving the calculated delta-function dispersion with broadening functions (Gaussian is default). 
Whilst the internal fitting routines of Horace (including for resolution convolution) requires type 1 model functions (yielding ![S(\mathbb{Q})]),
many actual user models are type 2, including `SpinW` and `Euphonic`.
Thus, the new PACE implementation should allow both types of models.
We also note, that when instrumental resolution convolution (`TobyFit`) is used,
the broadening function in `disp2sqw` should only represent some intrinsic physical broadening
(such as a quasiparticle lifetime).
Whereas when `TobyFit` is not used,
the broadening function in `disp2sqw` will be a combination of both intrinsic and instrumental resolution broadening.

In the current Horace implementation, these model functions are used by the `multifit_sqw` function/class
(which inherits from `multifit`) to fit to data using a built-in Levenberg-Marquardt optimizer. 

> The [new design for PACE](Model_Optimisation_Design.md) proposes instead to use a pair of classes, `OptFunction` and `OptModel`.

Reviewer: ```suggestion
The [new design for PACE](Model_Optimisation_Design.md) proposes instead to use a trio of classes, `OptFunction`, `OptFunctions`, and `OptModel`.
```
or since `OptFunctions` probably isn't relevant for this discussion 
```suggestion
The [new design for PACE](Model_Optimisation_Design.md) proposes instead to use a host of classes including `OptFunction` and `OptModel`.
```

`OptModel` will be more like the current `multifit` (or `tobyfit`) class,
whilst `OptFunction` is a wrapper around the user supplied model function to handle parameter bindings and bounds currently handled by `multifit`.
`OptFunction` should therefore have functionality to transform type 2 input to type 1 input

> whilst `OptModel` would expect only type 1 input.

Reviewer: I agree that `OptFunction` must contain the ability to transform function **output** into the equivalent of type 1.
In the case of resolution convolution the class inheriting from `OptFunction` and implementing the Tobyfit method would do this conversion at the start of its scope, but other convolution methods might perform the energy-axis convolution analytically in which case type 2 (or my suggested type 3) function output would be converted to type 1 as part of the convolution.

Author: Ok. I've changed the text to note that for standard transformations listed in the optimisation document that `OptFunction` will automatically do the transformation of both input and output, and for others, the user is responsible for the transformation.

However, for concreteness, *the rest of this document will assume that the new optimisation **framework** will be written in Matlab.*
The proposed API may need to change should it be decided to implement the optimisation framework in a different language.

# <a href="external_interfaces"></a> Matlab interfaces to external languages

One immediate restriction on non-Matlab user-defined model functions is that its *input* arguments must be "plain-old-data" (POD, numeric types or strings) arrays only, and not objects.
This is because Matlab does not allow its own objects (instances of Matlab-defined classes) to be passed to external libraries.
(A workaround for this for the Python user interface using a compiled Horace library exists,
and is described [here](../../python_interface/design/01_pace_python_high_level_discussion.md#interpreter-calls)).
This means that the second type of function signatures described in the [optimisation design document](Model_Optimisation_Design.md#optimisation-function-object), 
where Matlab objects (in particular an `sqw` object) are passed as input to the user defined function,

> will not be possible with non-Matlab functions, unless `OptFunction` unwraps this object into a (several) POD array(s).

Reviewer: `OptFunction` was designed with exactly this in mind. Wrapped functions will need to have their expected input specified (either explicitly or via some sort of function inspection) and `OptFunction` will contain the mechanism required to produce POD arrays when necessary.

Author: Sorry I was not clear. I meant that the user-defined function signature: 
> `arrays_objects(a, ..., p)` of which `S_of_Q_w(sqw_obj, p)` is a specific case

Will not be possible for non-Matlab defined `array_objects` or `S_of_Q_w` because `a` or `sqw` cannot be passed from Matlab to an external library.

Method 3 (inference from function name) will probably work for all languages,
but may prove difficult to implement for the compiled languages.
Method 4 would only work for Python and C++.

As discussed in the section on [user models written in compiled languages](#compiled) below,
Matlab has two ways to call functions in a compiled library,
the simpler of which uses a C interface (not C++).
*As noted there, we propose to use this C-style approach, in order to present a simple and uniform API for all compiled languages.*
However, this means that methods 2 and 4 for inferring input arguments are not feasible,
and method 3 would be difficult to implement as discussed in the [section on the C API](#c_functions).

## <a name="python"></a> Python

There are two use cases for user defined model functions written in Python:

1. The user is running Matlab (has a license) and wants to call a model function (e.g. Euphonic) written in Python.
2. The user is running Horace in Python using the compiled library and wants to call their own Python model function.

Taking the first case, it is currently possible to do this in Matlab

> (as exemplified by the [`meuphonic` project](https://github.com/pace-neutrons/Meuphonic)).

Reviewer: `meuphonic` has been renamed [`horace-euphonic-interface`](https://github.com/pace-neutrons/horace-euphonic-interface), however the old link does and will forever continue to work.

This requires, however, conversions between Matlab and Python data types,
(as demonstrated by the `m2p` and `p2m` methods of [Brille](https://github.com/g5t/brille/)),
which implicitly requires a memory copy in each direction
(one memory copy of each input and output arrays for a single function evaluation).

We can avoid the memory copy for the input argument arrays as long as they are restricted to be numeric arrays (`numpy` on Python-side) only.
This would use a `mex` file to wrap the Python function call and its arguments in a `numpy` array,
similar to the work described [here](../../python_interface/design/02_pace_python_implementation_discussion.md#pymatpy).

> As noted in the [Appendix](#app_nativememory), a memory copy of the output argument is only possible

Reviewer: ```suggestion
As noted in the [Appendix](#app_nativememory), avoiding a memory copy of the output argument is only possible
```
Is this missing a word?

then this header file must be generated dynamically.

The actual function signatures and API are discussed for each language separately in the sections below, 
but we should note here a limitation with using `loadlibrary` (and using dynamically loaded libraries in general).
This is because the C-language function may only accept pointers to arrays,
and because, as noted [in the appendix](app_nativememory) the main Matlab code and library codes have different memory heaps,
so they must each be responsible for allocating *and deallocating* their own memory.
That is, a memory error will occur if Matlab tries to deallocate a memory block allocated by the library and vice versa.
To avoid this, *we propose to not allow user functions to allocate their own memory for the results array*.
That is, Matlab will allocate an array for the results and pass the pointer to this to the user model function.
The user model function must then fill this array with the result of the calculations,
rather than using their own allocated array.

Finally we note that the compiled library interface deal *exclusively* with real-valued variables.

> The user defined model function is supplied with real input coordinates ![\mathbb{Q}] and is expected to return real ![S(\mathbb{Q})].

Reviewer 2: `![\mathbb{Q}]` is being rendered raw

It should thus not return complex values, even if the calculations result in complex values.
In these cases it should either raise an error or a warning.

# User defined model function signatures

## <a name="matlab_functions"></a> Matlab

```matlab
function user_model_sqw(qh, qk, ql, en, parameters, varargin)
function user_model_dsp(qh, qk, ql, parameters, varargin)
function user_model_pow(modQ, en, parameters, varargin)
function user_model_1d(en, parameters, varargin)
function user_model_0d(parameters, varargin)

> ```

Reviewer: What about the cases outlined in `OptFunction` where, e.g., plotting axes are passed to the function?
I propose that all prototypical function signatures be of the form `function user_model_NdX`? where `N` is how many equal-sized 1-D arrays are expected before the parameters and `X` indicates the return type listed above.

Author: I think the users can name their functions whatever they like... the names are just to make explicit what physical type of data the mappings would correspond to... the users would have to specify which case is which in the `OptFunction` constructor below.

I'm assuming that `X` would always be a double.

> This follows the convention established by the current implementation of `multifit`.

Reviewer 2: Is it easier to pass 3 or 4 `1xn` arrays of values for `qh`, `qk`, `ql`, `en` or a single `3xn` or `4xn` array?

Author: For Matlab and Python user functions, probably it doesn't matter either way. For the compiled library there might be an ambiguity on the data layout - is the Matlab 2D array passed column or row-major? Newer Matlab supports C-style row-major matrices but the ones I've tested all return a column-major array.

Also, unless we add another `int` to the function signature (in addition to `n_elem`) to determine the size of the input array, we would need to trust that the users will handle the input array size correctly. With multiple 1D input arrays we can use the runtime signatures checking in `calllib` to ensure that the correct library function is called.

So, on balance I would prefer to stick with multiple 1D arrays, unless it would be much more performant for Horace to pass a 2-D array.

```

etc. could be used instead, since `pyHorace` will have Python wrappers which can interpret "(string, param)" pairs in a cell array as keyword arguments.
To set up the `OptFunction` objects in `pyHorace` either keyword or positional arguments will be accepted, but not a mix:

```python
optfun_sqw = m.OptFunction(user_model_sqw, 'type', 'double', 'inputs', 4, 'parameters', parvec, 'arguments', arglist)
optfun_sqw = m.OptFunction(user_model_sqw, type='double', inputs=4, parameters=parvec, **kwargs_for_func)
```

A mix of keywords and Matlab-style pairs of keyword and positional arguments will give an error, as this mix can cause ambiguities which are hard to parse.

## <a name="c_functions"></a> C

As we are using `loadlibrary` / `calllib` to get Matlab to evaluate a function in a compiled library,
we need to use C-calling conventions for these functions for all languages: C, C++ and Fortran.
In order to accommodate Fortran, these functions *should pass by reference only*, and have no return values.

The function signatures in C is thus:

```c
void user_model_sqw(const double *qh, const double *qk, const double *ql, const double *en,

>                     const double *parameters, double *results, int *n_elem);

Reviewer: ```suggestion
                    const double *parameters, double *results, unsigned long *n_elem);
```
`n_elem` might be *very* large, it should be at least an unsigned 32-bit integer (2<sup>32</sup>-1 ~ 4.29 billion).

Also, similar to above, we should define signatures for the N<=4 dimensional user functions of type 1.

void user_model_dsp(const double *qh, const double *qk, const double *ql,
                    const double *parameters, double *result_omega, double *results_S, int *n_elem);
```

> where `n_elem` is the number of elements which Matlab expects `results` to have.

Reviewer: ```suggestion
where `n_elem` is the number of elements which Matlab expects `results` to have and the length of all input vectors.
```

a heap memory error will occur which will crash the program.
Matlab only supports the use of `createArrayFromBuffer` using Matlab allocated memory using the `createBuffer` method.
It is possible to modify the internals of a Matlab array to manually disassociate such an array from its buffer
and then safely delete the array but this requires extensive wrapping and is still a fragile design.

We have thus decided not to implement such a design and to either require functions
to fill in a Matlab allocated array or to live with a memory copy.

<!---
LaTeX equations used repeatedly can be defined as references, with format
	[latex_equation]: %-encoded link to image generator
then, if image generation fails for some reason, the alt-text for each equation
will still contain the raw LaTeX equation which is often intelligible

Constructing the %-encoded link can be done at, e.g., the codecogs website
	https://www.codecogs.com/eqnedit.php?latex=latex_equation
--->
[\mathbb{Q} \equiv (\mathb{Q},E)]:http://latex.codecogs.com/svg.latex?%5Cmathbb%7BQ%7D%5Cequiv%28%5Cmathbf%7BQ%7D%2CE%29
[S(\mathbb{Q})]: http://latex.codecogs.com/svg.latex?S%28%5Cmathbb%7BQ%7D%29

> [\mathbf{Q}]: http://latex.codecogs.com/svg.latex?%5Cmathbf%7BQ%7D

Reviewer: ```suggestion
[\mathbf{Q}]: http://latex.codecogs.com/svg.latex?%5Cmathbf%7BQ%7D
[\mathbb{Q}]: http://latex.codecogs.com/svg.latex?%5Cmathbb%7BQ%7D
```
To address the issue pointed out by Nick above.
